Index: samples/CMakeLists.txt
===================================================================
--- samples/CMakeLists.txt	(revision 521)
+++ samples/CMakeLists.txt	(working copy)
@@ -27,8 +27,8 @@
 add_executable(samples_cpp ${LIBTCOD_SAMPLE_CPP_SRC})
 target_link_libraries(samples_cpp ${LIBTCOD_NAME} ${LIBTCOD_CPP_NAME})  #${LIBTCOD_CPP_NAME} is empty on non-linux platforms
 
-add_executable(hmtool ${LIBTCOD_HP_TOOL_SRC})
-target_link_libraries(hmtool ${LIBTCOD_NAME} ${LIBTCOD_GUI_NAME}) 
+add_executable(hmtool/hmtool ${LIBTCOD_HP_TOOL_SRC})
+target_link_libraries(hmtool/hmtool ${LIBTCOD_NAME} ${LIBTCOD_GUI_NAME}) 
 
 IF(UNIX)
 	SET(CMAKE_EXE_LINKER_FLAGS "-Wl,-rpath,. ${CMAKE_EXE_LINKER_FLAGS}")
@@ -36,7 +36,7 @@
 
 set_property(TARGET samples_c PROPERTY DEBUG_POSTFIX "_debug")
 set_property(TARGET samples_cpp PROPERTY DEBUG_POSTFIX "_debug")
-set_property(TARGET hmtool PROPERTY DEBUG_POSTFIX "_debug")
+set_property(TARGET hmtool/hmtool PROPERTY DEBUG_POSTFIX "_debug")
 
 IF(CMAKE_BUILD_TYPE MATCHES "Release")
 	IF(NOT APPLE)
@@ -55,15 +55,15 @@
 		)
 	
 		ADD_CUSTOM_COMMAND(
-   		    TARGET hmtool
+   		    TARGET hmtool/hmtool
 		    POST_BUILD
-		    COMMAND ${UPX} hmtool${PROGRAM_ENDING}
+		    COMMAND ${UPX} hmtool/hmtool${PROGRAM_ENDING}
 		    VERBATIM
 		)
 	ENDIF()
 ENDIF()
 
-INSTALL(TARGETS samples_c samples_cpp hmtool
+INSTALL(TARGETS samples_c samples_cpp hmtool/hmtool
   RUNTIME DESTINATION ${INSTALL_DIR}
   LIBRARY DESTINATION ${INSTALL_DIR}
 )
